src/Interpreters/TreeRewriter.cpp

```diff
@@ -455,49 +462,34 @@ void removeUnneededColumnsFromSelectClause(ASTSelectQuery * select_query, const
     else
         return;
 
-    ASTs new_elements;
-    new_elements.reserve(elements.size());
+    ASTs new_elements(elements.size() + output_columns_positions.size());
+    size_t new_elements_size = 0;
 
     NameSet remove_columns;
 
-    /// Resort columns according to required_result_columns.
-    if (reorder_columns_as_required_header && !required_result_columns.empty())
-    {
-        std::unordered_map<String, size_t> name_pos;
-        {
-            size_t pos = 0;
-            for (const auto & name : required_result_columns)
-                name_pos[name] = pos++;
-        }
-        ::sort(elements.begin(), elements.end(), [&](const auto & lhs, const auto & rhs)
-        {
-            String lhs_name = lhs->getAliasOrColumnName();
-            String rhs_name = rhs->getAliasOrColumnName();
-            size_t lhs_pos = name_pos.size();
-            size_t rhs_pos = name_pos.size();
-            if (auto it = name_pos.find(lhs_name); it != name_pos.end())
-                lhs_pos = it->second;
-            if (auto it = name_pos.find(rhs_name); it != name_pos.end())
-                rhs_pos = it->second;
-            return lhs_pos < rhs_pos;
-        });
-    }
-
     for (const auto & elem : elements)
     {
         String name = elem->getAliasOrColumnName();
+        size_t result_index = output_columns_positions.size() + new_elements_size;
+        if (auto it = output_columns_positions.find(name); it != output_columns_positions.end())
```

Comment:
As I understand, here we choose which one column we use in the result in the case, when there are several columns with the same name. But `std::miltimap::find` is not deterministic. This probably can be a problem.

Reply:
It should not be a problem, since columns should be the same, and it does not matter which one will be prefered.